File: cpp/DSA/stack/1_a_stack.cpp
#include<iostream>
using namespace std;

class Stack{
    int topIndex;
    int maximumSize;
    int* stack;

    public:
    Stack(int maxSize = 5){
        maximumSize = maxSize;
        stack= new int[maximumSize];
        topIndex = -1;   
    }

    void push(int elemnt){
        if(topIndex < maximumSize -1){
            stack[++topIndex]=elemnt;
        }
        else{
            cout<<"Stack is full";
        }
    }
    void top(){
        if(topIndex >= 0){
            cout << "Top element is : "<< stack[topIndex] << endl;
        }
        else{
            cout << "Stack is empty!" << endl;
        }
    }
    void pop(){
        if(topIndex >= 0){
            cout << "Removed element is : " << stack[topIndex--]<< endl;
        }
        else{
             cout << "Stack is empty!" << endl;
        }
    }
    };

int main(){
    Stack s;
    s.push(10);
    s.push(20);
    s.push(30);
    s.top();
    s.pop();
    s.top();
    return 0;
}

File: python/DSA/stack/1_a_stack.py
class Stack:
    def __init__(self, max_size=5):
        self.max_size = max_size
        self.stack = []
        
    def push(self, element):
        if len(self.stack) < self.max_size:
            self.stack.append(element)
            print(f"Inserted {element}")
        else:
            print(f"Stack is full. Max size is {self.max_size}")
    
    def top(self):
        if len(self.stack) > 0:
            print("Top element: ", self.stack[len(self.stack)-1])
        else:
            print("Stack is empty")
    
    def pop(self):
        if len(self.stack) > 0:
            print("removed element: ",  self.stack.pop())
        else:
            print("Stack is empty")
            return None



elements = [1, 2, 3, 4, 5, 6]

print("Stack with no Max Size passed")
s1 = Stack()
for element in elements:
    s1.push(element)

s1.top()  
s1.pop()  
s1.top()  


print("Stack with Max Size passed")
s2 = Stack(max_size=10)
for element in elements:
    s2.push(element)

s2.top()  
s2.pop()  
s2.top() 